ledger/src/blockstore.rs

```diff
@@ -2923,13 +2924,68 @@ impl Blockstore {
         self.last_root()
     }
 
+    pub fn lowest_cleanup_slot(&self) -> Slot {
+        *self.lowest_cleanup_slot.read().unwrap()
+    }
+
     pub fn storage_size(&self) -> Result<u64> {
         self.db.storage_size()
     }
 
     pub fn is_primary_access(&self) -> bool {
         self.db.is_primary_access()
     }
+
+    pub fn scan_and_fix_roots(&self, exit: &Arc<AtomicBool>) -> Result<()> {
+        if !self.is_root(self.last_root()) {
+            self.set_roots(&[self.last_root()])?;
+        }
+
+        let lowest_scan_slot = std::cmp::max(self.lowest_cleanup_slot(), self.lowest_slot());
+        let ancestor_iterator = AncestorIterator::new(self.last_root(), &self)
+            .take_while(|&slot| slot >= lowest_scan_slot);
+
+        let mut find_missing_roots = Measure::start("find_missing_roots");
+        let mut roots_to_fix = vec![];
+        for slot in ancestor_iterator.filter(|slot| !self.is_root(*slot)) {
```

Comment:
Might also be good to include a timer for this scan, similar to `Measure::start("fix_roots");` below